client/tinychain/ml/dnn.py

```diff
@@ -12,50 +12,179 @@ def layer(weights, bias, activation):
+            '''
             Z = einsum("ij,ki->kj", [self.weights, inputs])
             A = activation.forward(Z) + self.bias
             return A, Z
 
         def update(self, d_weights, d_bias):
+            '''
+            update function is used to ensure that update operations are done
+            New_Weight = Old Weight - Delta Weight
+            New_Bias = Old Bias - Delta Bias
+            '''
             return self.weights.write(self.weights - d_weights), self.bias.write(self.bias - d_bias)
+                
 
     return DNNLayer([weights, bias])
 
 
+def optimizers():
+    """Construct alternative optimizers for Gradient Descent."""
+    
+   class Adam_Optimizer(optimizers):
+       
+       @classmethod
+       def create(cls):
+            return tc.Map({"l_m":tc.Number(0) ," l_v":tc.Number(0) , "t":tc.Number(0)})
+        
+       def optimize (self, A_prev, dA, Z ,decay_rate_1 : tc.Number, decay_rate_2 : tc.Number,
+                     epsilon :tc.Number , l_m,l_v,t):  
+                             
+            dZ = activation.backward(dA, Z).copy()          
+            # Gradients for each layer
+            g = einsum("kj,ki->ij", [dZ, A_prev])              
+            t=t+1           
+            # Computing 1st and 2nd moment for each layer
+            l_m = l_m * decay_rate_1 + (1- decay_rate_1) * g            
+            l_v = l_v * decay_rate_2 + (1- decay_rate_2) * (g ** 2)            
+            l_m_corrected = l_m / (1-(decay_rate_1 ** t))
+            l_v_corrected = l_v / (1-(decay_rate_2 ** t))
+            
+            # Update Weights
+            d_weights = l_m_corrected / ((l_v_corrected)**(0.5) + epsilon)                       
+            d_bias = dZ.sum(0)
+            
+            return tc.Map({'l_m':l_m,
+                           'l_v':l_v,
+                           't':t,
+                           'd_weights': d_weights,
+                           'd_bias':d_bias})
+        
+        def update(self,d_weights ,d_bias)
+            
+            return self.weights.write(self.weights - d_weights), self.bias.write(self.bias - d_bias)
+
+
+
 def neural_net(layers):
     """Construct a new deep neural network with the given layers."""
 
     num_layers = len(layers)
 
     class DNN(NeuralNet):
         def eval(self, inputs):
-            state = layers[0].eval(inputs)
+            '''
+            eval function it will help you to get the result
+            of each layer
+            
+            for the first layer its exception we eval the inputs
+            
+            type(inputs) : tc.tensor.Dense
+            state : is the Result from evaluating  cascaded layers 
+            '''
+            state = layers[0].eval(inputs) 
             for i in range(1, len(layers)):
                 state = layers[i].eval(state)
 
             return state
+       
+
 
         def train(self, inputs, cost):
```

Comment:
I think this should accept an optimizer as an argument, then you can construct a new optimizer in the NeuralNetTests and actually test your code

Reply:
yes  ,i will add   ( ..., optimizer)